tests/agent_server/test.py

```diff
@@ -0,0 +1,293 @@
+
+            if len(actions) == 0:
+                after = None
+            else:
+                after = actions[-1].get("started")
+
+        return None
+
+    last_change_1 = await get_last_change(
+        resource_type="test::Resource",
+        agent="agent1",
+        key="key",
+        value="key1",
+    )
+
+    last_change_2 = await get_last_change(
+        resource_type="test::Resource",
+        agent="agent1",
+        key="key",
+        value="key2",
+    )
+
+    last_change_3 = await get_last_change(
+        resource_type="test::Resource",
+        agent="agent1",
+        key="key",
+        value="key3",
+    )
+
+    first_change_1 = await get_first_change(
+        resource_type="test::Resource",
+        agent="agent1",
+        key="key",
+        value="key1",
+    )
+
+    first_change_2 = await get_first_change(
+        resource_type="test::Resource",
+        agent="agent1",
+        key="key",
+        value="key2",
+    )
+
+    first_change_3 = await get_first_change(
+        resource_type="test::Resource",
+        agent="agent1",
+        key="key",
+        value="key3",
+    )
+
+    import json
+    logger.info("Last change for key1: " + json.dumps(last_change_1, indent=2))
+    logger.info("Last change for key2: " + json.dumps(last_change_2, indent=2))
+    logger.info("Last change for key3: " + json.dumps(last_change_3, indent=2))
+    logger.info("First change for key1: " + json.dumps(first_change_1, indent=2))
+    logger.info("First change for key2: " + json.dumps(first_change_2, indent=2))
+    logger.info("First change for key3: " + json.dumps(first_change_3, indent=2))
+
+
+    async def get_dependencies(id: str) -> str:
+        logger.info(id)
+        result = await client.get_resource(environment, id=id, logs=True)
+        assert result.code == 200
+
+        return result.result.get("resource", {}).get("attributes", {}).get("requires")
+
+
+    async def need_redeploy(
+        resource_type: str,
+        agent: str,
+        key: str,
+        value: str,
+    ) -> bool:
+        last_deployment = await get_last_change(resource_type, agent, key, value)
+        logger.info(f"Last deployment of self: {last_deployment['finished']}")
+
+        # Check dependencies
+        dependencies = await get_dependencies(
+            f"{resource_type}[{agent},{key}={value}],v={last_deployment.get('version')}"
+        )
+
+        resource_id = re.compile(r"(.*)\[(.*),(.*)=(.*)\]")
+
+        # For each dependency, check first change after last_deployment
+        # If any change, needs redeploy
+        for dependency in dependencies:
+            reg_match = resource_id.match(dependency)
+            dep_resource_type = reg_match.group(1)
+            dep_agent = reg_match.group(2)
+            dep_key = reg_match.group(3)
+            dep_value = reg_match.group(4)
+
+            logger.info(f"Checking dependency {dependency}")
+            first_change = await get_first_change(
```

Comment:
I in fact doesn't have to be the first change, but any change, so it could also re-use the last_change method

Reply:
There is only one method now